### ISSUE TITLE
No option to change text direction (RTL vs LTR) in version 3?

### ISSUE THREAD

**Issue by kavir1698** [Zettlr/Zettlr#4596]
### Checklist

- [X] Zettlr does not possess this feature (if it does, but the feature doesn't work out for you, propose an enhancement instead.)
- [X] Zettlr's current features cannot be used to approximate this behaviour (otherwise, propose an enhancement instead.)
- [X] I have searched the documentation to make sure I did not overlook it.
- [X] I have searched both open and closed issues to see if someone has already proposed this.

### Feature Proposal

There was an option in the editor preferences to change the text direction. I just updated to version three and the feature is not there, nor does it automatically choose the direction. Could you please bring this option back?

### Example Workflow

A toggle to change the text direction.

### Caveats

none

### Do you wish to attempt implementing this yourself?

No

### Platforms

- [X] Windows
- [ ] macOS
- [ ] Linux

**Comment by nathanlesage** [comment]
> nor does it automatically choose the direction

This is — allegedly — working out of the box, but I may have misread some part of CodeMirror's documentation. Feel free to play around with it to see what is needed to make the text direction actually change.

**Comment by mfakih294** [comment]
I confirm this issue. In v3, Arabic text and English text are left aligned. In v2.3.0 Arabic text used to be displayed right to left automatically. A sample Arabic text: 
هذا السطر ينبغي أن يظهر من يمين الشاشة.

**Comment by nathanlesage** [comment]
Okay, I just had another look at what CodeMirror does here and what I need to do. Turns out I just hadn't looked well enough. So here's what I have gathered:

1. CodeMirror relies as much as possible on CSS, and it turns out that setting `dir="rtl"` in the surrounding editor DIV makes the script go right-to-left (albeit it's an all-or-nothing thing).
2. There are a few methods that allow working with bidi, albeit I do not understand exactly where or how these are supposed to work.
3. Auto-detection does not appear to work (i.e. setting `dir="auto"`)
4. Fortunately, there is an example live on the website that wasn't there until a few weeks ago, so it seems we can make something properly work: https://codemirror.net/examples/bidi/
5. A related thread is [here](https://discuss.codemirror.net/t/overriding-bidi-spans-in-rtl-context/6695)

**So regarding a quick relief for RTL-writers here, I have a question**:

Would it be okay if I first provided only a very coarse direction-change setting to switch between RTL and LTR? That may mess up LTR-parts in the code, though, but would give us a first impression as to how that works and should at least improve the situation until we know more.

Or would you prefer to live with an entirely LTR-environment for now until I have properly figured out all the shenanigans with changing text direction? That might take even a few months because there is a lot to take care about and I have a full-time job …

What's your preference?

**Comment by kavir1698** [comment]
Thanks for the update. I would prefer a coarsely working version for now, as I use Zettlr heavily for my writings in Persian.

**Comment by nathanlesage** [comment]
Okay, then I'll quickly implement something like that. Would you mind providing me a test file that I can use to verify coarsely that it does what I intend it to?

Ideal would be a file that uses bidi-text (i.e. both RTL and LTR text) and formatting characters (links, images, etc.) to see if something really badly breaks or if it looks sufficiently usable. The text itself doesn't need to be important, it's mostly about characters from both scripts!

**Comment by mfakih294** [comment]
1 - In CSS, I use the property unicode-bidi: plaintext; (this keyword makes the elements directionality calculated without considering its parent bidirectional state or the value of the direction property. The directionality is calculated using the P2 and P3 rules of the Unicode Bidirectional Algorithm. This value allows the display of data that is already formatted using a tool following the Unicode Bidirectional Algorithm.) to enable auto direction.

2 - A common use case is to write in one language per file, or even one language for all files. For these cases, a setting to enable global direction is a good workaround for now.

3 - Provided below is a text file containing two scripts: English and Arabic:

[sample-ar-en.md](https://github.com/Zettlr/Zettlr/files/12698488/sample-ar-en.md)

**Comment by mfakih294** [comment]
And below is how it looked in Visual Studio Code with RTL markdown enabled (all right aligned).

[screenshot.zip](https://github.com/Zettlr/Zettlr/files/12698601/screenshot.zip)

Note: had to put in a zip file as I had problems upload image files (firewall issues).

**Comment by mfakih294** [comment]
Here's a relevent discussion from a [forum](https://forum.obsidian.md/t/mixed-rtl-and-ltr/34808/39?page=2)of an editor that uses CodeMirror (Obsidian).

**Comment by nathanlesage** [comment]
Good point, this plugin here looks comprehensive and capable of being adapted for our use case: https://github.com/esm7/obsidian-rtl/pull/48/files

**Comment by zoli** [comment]
Hi @nathanlesage,
Just wanted to mention that in the new version MIT license is added to the project. As you said before the implementation in Zettlr would be much the same. Let me know if there is any ambiguity or problem.

**Comment by nathanlesage** [comment]
@zoli Wonderful, thank you for pinging me. I just had a look, and with a few months in between thinking about it, the plugin makes a ton of sense, and should be easy to adapt.

Since I just saw that it's your PR: If you want to attach your name to it in Zettlr as well, you can open a PR implementing this for the app. Otherwise, I'd do a copy & paste job, but then your GH username won't be in the commits. How do you prefer? Should you choose to implement this, I'd quickly think about the most efficient and easiest way of doing so for you & me so that you have an easy list of to-dos (& my support in case of any obstacles).

**Comment by zoli** [comment]
No problem about the name, you can use the code and add the feature.

**Comment by mhawash** [comment]
I would suggest making the RTL feature has the option to be dynamic/bidirectional, based on each line maybe.

**Comment by alr86** [comment]
any update?

**Comment by mhawash** [comment]
We're still waiting for this!
